### PR TITLE
Add 'Network error handling' section to documentation

### PR DESCRIPTION
**Pull request by phpbg**
### Description

Document network errors handling and reconnect strategy

---

### Checklist

- [x] Does `npm test` pass with this change (including linting)?
- [x] Is the new or changed code fully tested?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?

### CODE REVIEW

**Comment by leibale** [comment]
I think it'll fit better in the [client-configuration.md](https://github.com/redis/node-redis/blob/master/docs/client-configuration.md) file

**Comment by phpbg** [comment]
i didn't even notice there was more doc... lol

I guess it isn't relevant anymore?

**Comment by leibale** [comment]
@phpbg I think that we should merge your changes into that doc somehow.. :)

**Comment by phpbg** [comment]
ok, i'll move them to [client-configuration.md](https://github.com/redis/node-redis/blob/master/docs/client-configuration.md)

**Comment by phpbg** [comment]
@leibale It should be in better shape now ;-)

File: docs/client-configuration.md [hunk 1]
```diff
@@ -28,13 +28,30 @@
 
 ## Reconnect Strategy
 
-You can implement a custom reconnect strategy as a function:
+When a network error occurs the client will automatically try to reconnect, following a default linear strategy (the more attempts, the more waiting before trying to reconnect).
+
+This strategy can be overriden by providing a `socket.reconnectStrategy` option during client's creation.
+
+The `socket.reconnectStrategy` is a function:
```
Comment (chayim) [inline docs/client-configuration.md:35]:
```
The `socket.reconnectStrategy` is a function that:
```

**Review by chayim — APPROVED** [review]
One small change to the text, approving.

File: docs/client-configuration.md (hunk 1 above)
Comment (phpbg) [reply docs/client-configuration.md]: Fixed

File: README.md [hunk 2]
```diff
@@ -46,7 +46,10 @@ import { createClient } from 'redis';
 
 const client = createClient();
 
-client.on('error', (err) => console.log('Redis Client Error', err));
+client.on('error', (err) => {
+    // Do **NOT** throw here as it would break auto reconnect
```
Comment (leibale) [inline README.md:50]: I'm actually not too sure if that's true (and if it is then its a bug IMO)

File: docs/client-configuration.md [hunk 3]
```diff
@@ -28,13 +28,30 @@
 
 ## Reconnect Strategy
 
-You can implement a custom reconnect strategy as a function:
+When a network error occurs the client will automatically try to reconnect, following a default linear strategy (the more attempts, the more waiting before trying to reconnect).
+
+This strategy can be overriden by providing a `socket.reconnectStrategy` option during client's creation.
```
Comment (leibale) [inline docs/client-configuration.md:33]: 1. overridden
2. during **the** client's creation.

File: docs/client-configuration.md [hunk 4]
```diff
@@ -28,13 +28,30 @@
...
     - `number`: wait time in milliseconds prior to attempting a reconnect.
     - `Error`: closes the client and flushes internal command queues.
 
+The example below shows the default `reconnectStrategy` and how to override it.
+
+```typescript
+import { createClient } from 'redis';
+
+const client = createClient({
+    url: 'redis://alice:foobared@awesome.redis.server:6380',
```
Comment (leibale) [inline docs/client-configuration.md:48]: no need to show url in this example

**Review by leibale — CHANGES_REQUESTED** [review]

**Comment by leibale** [comment]
@phpbg sorry for the long dealy, I was on vacation..

File: README.md (hunk 2 above)
Comment (phpbg) [reply README.md:50]: Hi @leibale 
no one will catch if you thow here. I (painfully) experienced it :-D

I think you should generally never throw in an event handler.

If the doc makes it clear, then this is not a bug.

This is the reason why I opened this PR.

Comment (leibale) [reply README.md]: this bug is fixed

File: docs/client-configuration.md (hunk 3 above)
Comment (phpbg) [reply docs/client-configuration.md]: fixed

File: docs/client-configuration.md (hunk 4 above)
Comment (phpbg) [reply docs/client-configuration.md]: fixed

File: README.md (hunk 2 above)
Comment (phpbg) [reply README.md]: Good to hear ;-)

Out of curiosity, do you have a commit id? I wonder how it was fixed...

Anyway the comment was removed. PR is ready.

**Comment by leibale** [comment]
Blocked by #2373 - will need some updates as this PR changes `reconnectStrategy` a bit:
1. It now can be set to `false` to  disable reconnect (instead of a function that always returns an error)
2. The function now gets a new parameter - `cause: unknown` - which is the thing that was thrown and triggered reconnect (the type is `unknown` and not `Error` because anything is `throw`able in JS).

**Comment by leibale** [comment]
Merging as is, will make the changes in #2373

**Pull request merged**